### PR TITLE
[DEV APPROVED] 9827 First time buyer journey for Scottish property buyers

### PR DESCRIPTION
**Pull request by margOnline**
[TP 9827](https://moneyadviceservice.tpondemand.com/restui/board.aspx?#page=board/5682182231901749200&appConfig=eyJhY2lkIjoiRkRBMzk3RTBGNjU3NDg1RDM4QzczMTJFNzI5MDFBN0MifQ==&boardPopup=userstory/9827/silent)

### Requirement
Currently, the calculations for property buying in Scotland address 2 journeys:
screen shot 2019-02-25 at 13 11 40

This feature caters for a 3rd user journey, the first time buyer
screen shot 2019-02-25 at 13 12 06

### Approach
To understand the approach and the changes, it may help to refer to pr #332 and pr #333.

#### Abbreviations
SDLT - Stamp Duty Land Tax, payable in England and Northern Ireland on property purchases
LBTT - Land and Buildings Transaction Tax, Scottish version of SDLT
LTT - Land Transaction Tax, Welsh version of SDLT
FTB - First Time Buyer

Because this feature makes the user journey for Scottish property buyers more akin to property buyers in England than property buyers in Wales, I have approached this feature as follows:

#### tldr;
Previously LBTT and LTT inherited from the same parent controller and used shared views; this feature changes the inheritance structure so that LBTT and SDLT inherit from the same parent controller and LTT doesn't. Refer to the commit messages for details.

#### Technical
1. Decouple LBTT calculator controller from the parent Transaction Tax calculator controller
controller which was originally set up for the Scottish and Welsh transaction tax calculators.
2. Delete the LTT calculator's parent controller because inheriting from a parent is now redundant.
3. Extract the similarities between the Scottish and English journeys to a
parent controller using Rails's `single template inheritance` to avoid duplication in views.

### CODE REVIEW

**Comment by masbuild** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          7 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>app/controllers/mortgage_calculator/land_and_buildings_transaction_taxes_controller.rb</td>
      <td>17</td>
      <td>Layout/EmptyLines: Extra blank line detected.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>app/controllers/mortgage_calculator/land_transaction_taxes_controller.rb</td>
      <td>15</td>
      <td>Style/GuardClause: Use a guard clause instead of wrapping the code inside a conditional expression.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>app/controllers/mortgage_calculator/land_transaction_taxes_controller.rb</td>
      <td>15</td>
      <td>Style/IfUnlessModifier: Favor modifier <code>unless</code> usage when having a single-line body. Another good alternative is the usage of control flow <code>&amp;&amp;</code>/<code>||</code>.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>app/helpers/mortgage_calculator/land_transaction_taxes_helper.rb</td>
      <td>25</td>
      <td>Style/TrailingCommaInHashLiteral: Avoid comma after the last item of a hash.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/helpers/land_and_buildings_transaction_taxes_helper_spec.rb</td>
      <td>19</td>
      <td>RSpec/ContextWording: Start context description with ‘when’, ‘with’, or ‘without’.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/helpers/land_transaction_taxes_helper_spec.rb</td>
      <td>19</td>
      <td>RSpec/ContextWording: Start context description with ‘when’, ‘with’, or ‘without’.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false"><table>
  <tbody>
    <tr>
      <td>spec/models/land_and_buildings_transaction_tax_spec.rb</td>
      <td>341</td>
      <td>Layout/EmptyLines: Extra blank line detected.</td>
    </tr>
  </tbody>
</table>
</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is huge. Consider next time to split up the workinto multiple PRs. A good guide about small PRs can be seenhere(link to mas-standards - small PRs guide)</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">It looks like this is a Rails engine, but no changes to <a href="https://github.com/moneyadviceservice/mortgage_calculator/blob/7b6e2029c57a8098b7d04ea27abcd2e26c4523ad/lib/mortgage_calculator/version.rb">lib/mortgage_calculator/version.rb</a> detected. Did you forget to bump the version?</td>
    </tr>
  </tbody>
</table>

At Money Advice Service, we use Danger gem, to ensure that our commits and pull requests follow [standards](https://github.com/moneyadviceservice/mas-standards/blob/master/Dangerfile). If you think one of the points here has been raised in error (or shouldn't be in our standards), open an issue or pull request on [our Danger config](https://github.com/moneyadviceservice/mas-standards).

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: app/views/mortgage_calculator/land_and_buildings_transaction_taxes/_first_time_buyer_explanation.html.erb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+<p><%= I18n.t('land_and_buildings_transaction_tax.how_calculated_ftb_5', amount: ftb_threshold) %></p>
+<ul class="list stamp-duty__calculation-explanation-list">
+  <% I18n.t('land_and_buildings_transaction_tax.how_calculated_ftb_default_list').each do |default_list| %>
+    <li><%= default_list %></li>
+  <% end %>
+</ul>
+<p><%= I18n.t('land_and_buildings_transaction_tax.how_calculated_ftb_6') %></p>
+<p><%= I18n.t('land_and_buildings_transaction_tax.how_calculated_ftb_7') %></p>
+
+<%= render 'bands_table', rates: @rates %>
```
Comment (malcolmbaig) [inline app/views/mortgage_calculator/land_and_buildings_transaction_taxes/_first_time_buyer_explanation.html.erb]: It's good practice to refrain from using instance variables in partials, passing in local vars instead at the point of rendering. It helps make it clear at the top-level what the dependencies are for a given template to render successfully, without having to dig through each child partial.

**Comment by malcolmbaig** [comment]
Looking pretty good. I do wonder if accepting the duplication in two areas of the codebase might be clearer long-term than introducing the additional complexity of inheritance. I don't think it makes a huge difference either way, so LGTM 👍 .

File: app/views/mortgage_calculator/property_tax_calculator/create.html.erb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<div class="stamp-duty stamp-duty__container">
+  <h1 class="stamp-duty__heading"><%= I18n.t("#{i18n_locale_namespace}.heading_results") %></h1>
+  <%= render 'form_step2' %>
+  <div>
```
Comment (davidtrussler) [inline app/views/mortgage_calculator/property_tax_calculator/create.html.erb]: This looks like a surplus container.

Comment (margOnline) [reply app/views/mortgage_calculator/property_tax_calculator/create.html.erb]: good spot @davidtrussler ♻️

File: app/views/mortgage_calculator/land_and_buildings_transaction_taxes/_first_time_buyer_explanation.html.erb (hunk 1 above)
Comment (margOnline) [reply app/views/mortgage_calculator/land_and_buildings_transaction_taxes/_first_time_buyer_explanation.html.erb]: @malcolmbaig It seemed there was an inconsistent usage of passing locals to partials. I have standardised on doing so for the 3 'property tax' calculators. ♻️

**Comment by margOnline** [comment]
@malcolmbaig re your comment about introducing additional complexity to reduce duplication: it may be worth bearing in mind that there are 3 property tax calculators. I suspect the fact that there were 3 sets of very similar views, controllers and helpers may have motivated the original decision, taken in pr #332, to introduce inheritance. @tomas-stefano may have more to say on that decision. 

The use of inheritance could be extended to the 3rd property tax calculator to further reduce duplication. Happy to open up that question for comments.

File: app/views/mortgage_calculator/property_tax_calculator/_next_home_buyer_explanation.html.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<p><%= I18n.t("#{i18n_locale_namespace}.how_calculated") %></p>
+<p><%= I18n.t("#{i18n_locale_namespace}.how_calculated_additional") %></p>
+
+<%= render 'bands_table', rates: @rates %>
```
Comment (MarcoBgn) [inline app/views/mortgage_calculator/property_tax_calculator/_next_home_buyer_explanation.html.erb]: Referring to @malcolmbaig 's observation, shouldn't we pass locals up to here as well?

**Comment by tomas-stefano** [comment]
I think the whole situation it was how the things were implemented.

1) First everything was duplicated - the views, helpers and the controllers were all duplicated. And on this duplication there was nothing that was different between controllers/views/helpers.
2) Second it was implemented using template inheritance from Rails to remove duplication everywhere.
3) and now the solution of this PR. 

I think it is not ideal the way it is but it is definitely better and evolving IMO.

File: app/views/mortgage_calculator/property_tax_calculator/_next_home_buyer_explanation.html.erb (hunk 3 above)
Comment (margOnline) [reply app/views/mortgage_calculator/property_tax_calculator/_next_home_buyer_explanation.html.erb]: @MarcoBgn ♻️

Comment (MarcoBgn) [reply app/views/mortgage_calculator/property_tax_calculator/_next_home_buyer_explanation.html.erb]: Thanks @margOnline
